server/api.ts

```diff
@@ -552,29 +552,35 @@ router.postAsync("/match", ensure.isAdmin, async (req, res) => {
   res.send(match);
 });
 
+
+type SubmitWarmupBody = {
+  match: string; // Match ID
+  playerNo: 1 | 2; // Player number in match (1 or 2)
+  warmup: string; // Warmup map link or ID
+  mod: WarmupMod; // Mod used for warmup (only NM or DT is supported, defaults to NM)
+}
 /**
  * POST /api/warmup
  * Submit a warmup
- * Params:
- *   - match: match ID
- *   - playerNo: 1 - player 1, 2 - player 2
- *   - warmup: the warmup map, could be beatmap link or ID
  */
-router.postAsync("/warmup", ensure.loggedIn, async (req, res) => {
+router.postAsync("/warmup", ensure.loggedIn, async (req: Request<{}, SubmitWarmupBody>, res) => {
   const match = await Match.findOne({ _id: req.body.match }).orFail();
   if (!(await canEditWarmup(req.user, req.body.playerNo, match))) {
     logger.warn(
-      `${req.user.username} tried to submit player ${req.body.playerNo} warmup for ${req.body.match}`
+      `${req.user?.username} tried to submit player ${req.body.playerNo} warmup for ${req.body.match}`
     );
     return res.status(403).send("You don't have permission to do that");
   }
   try {
-    match[`warmup${req.body.playerNo}`] = await parseWarmup(req.body.warmup, match.tourney);
+    // Make sure mod is valid, if it's not default to nm
+    if (req.body.mod != "DT") { req.body.mod = "NM" }
+    match[`warmup${req.body.playerNo}`] = await parseWarmup(req.body.warmup, req.body.mod || "NM", match.tourney);
+    match[`warmup${req.body.playerNo}Mod`] = req.body.mod;
   } catch (e) {
     res.status(400).send(e.message);
     return;
   }
-  logger.info(`${req.user.username} submitted warmup ${req.body.warmup}`);
+  logger.info(`${req.user?.username} submitted warmup ${req.body.warmup}`);
```

Comment:
btw there is also this helper from util.ts you can use if you don't want to keep checking user for undefined
```
const user = assertUser(req);
```

Reply:
Ahh didn't know, applied in cb935bf6da1ca23fe77b22eff90b5ea1215b2199

Though I kept the permission check functions to allow undefined user being passed in